system/datatypes/mailinglist.php

```diff
@@ -412,246 +439,344 @@ private function unsubscribeAction() {
-			notify('error', __('ml-invalid-command'));
-
+		// return false if status is draft without client logged in
+		if (self::MAILING_STATUS_DRAFT == $mailing->getAttribute('status') && !isUser()) {
+			notify('error', __('not-found')); // 404
 			return null;
 		}
-		$isPosted = $saveMailing;
-
-		if (isset($_POST[self::FIELD_NAME_ID])) {
-			if (null == $mailingId) {
-				$mailingId = $_POST[self::FIELD_NAME_ID];
-			} elseif ($_POST[self::FIELD_NAME_ID] != $mailingId) {
-				notify('error', __('ml-inconsistent-parameters'));
 
-				return null;
+		// display mailing
+		/** @var HyphaDomElement $main */
+		$main = $this->html->find('#main');
+		$main->append('<div><h2>' . $mailing->getAttribute('subject') . '</h2></div>');
+		$date = $mailing->getAttribute('date');
+		if ($date) {
+			$date = new DateTime($date);
+			$main->append('<div>' . __('date') . ': ' . $date->format(__('ml-date-format')) . '</h2></div>');
+		}
+		if (isUser()) {
+			$receivers = $mailing->getAttribute('receivers');
+			if (null != $receivers) {
+				$main->append('<div>' . __('ml-received-by') . ': ' . $receivers . '</div>');
 			}
 		}
+		$main->append($mailing->getHtml());
 
-		// check given id
-		/** @var HyphaDomElement $mailing */
-		if (null != $mailingId) {
-			$mailing = $this->xml->document()->getElementById($mailingId);
-			if (!$mailing instanceof HyphaDomElement) {
-				notify('error', __('not-found')); // 404
+		// buttons
+		/** @var HyphaDomElement $commands */
+		$commands = $this->html->find('#pageCommands');
+		$commands->append($this->makeActionButton(__('back')));
 
-				return null;
-			}
-			// check if status is correct
-			$status = $mailing->getAttribute('status');
-			if (self::MAILING_STATUS_SENT == $status) {
-				notify('success', ucfirst(__('ml-successfully-sent')));
-				return ['redirect', $this->constructFullPath($this->pagename . '/' . $mailingId)];
-			}
-			if (self::MAILING_STATUS_DRAFT != $status) {
-				notify('error', __('unable-to-edit'));
+		// get status
+		$status = $mailing->getAttribute('status');
 
-				return null;
-			}
+		// add edit button when in draft
+		if (self::MAILING_STATUS_DRAFT == $status) {
+			$path = str_replace('{id}', $mailing->getId(), self::PATH_MAILS_EDIT_ID);
+			$commands->append($this->makeActionButton(__('edit'), $path));
+			$xpath = '//' . self::FIELD_NAME_ADDRESS . '[@' . self::FIELD_NAME_STATUS . '="' . self::ADDRESS_STATUS_CONFIRMED . '"]';
+			$num = count($this->xml->findXPath($xpath));
+
+			$path = str_replace('{id}', $mailing->getId(), self::PATH_MAILS_VIEW_ID);
+			$path = $this->language . '/' . $this->pagename . '/' . $path;
+			$commands->append(makeButton(__('send'), 'if(confirm(\'' . __('ml-sure-to-send', array("count" => $num)) . '\'))' . makeAction($path, self::CMD_SEND, '')));
+			$commands->append(makeButton(__('ml-test-send'), 'hypha(\''.$path.'\', \''.self::CMD_SEND_TEST.'\', prompt(\'' . __('email') . '\'), $(this).closest(\'form\'));'));
```

Comment:
I would use `json_encode()` for all these javascript strings.

Reply:
It was already like this and take too much time to implement, so I consider this outside of the scope of this pull request.

Reply:
Why would this take time to implement? I meant this:

```
$commands->append(makeButton(__('send'), 'if(confirm(' . json_encode(__('ml-sure-to-send', array("count" => $num))) . '))' . makeAction($path, self::CMD_SEND, '')));
$commands->append(makeButton(__('ml-test-send'), 'hypha('.json_encode($path).', '.json_encode(self::CMD_SEND_TEST).', prompt(' . json_encode(__('email')) . '), $(this).closest(\'form\'));'));
```

Reply:
... So it is fixed now.. by also replacing `"` by `'`

Reply:
Huh? Why is replacing " by ' needed? Javascript strings can be enclosed in " just fine, right? And `json_encode` will have taken care of any needed escaping already? If you replace the quote type, and your encoded strings contain quotes, I think you will actually break things?